### ISSUE TITLE
Unable to reproduce `GeoMath` and `Geometry3k`

### ISSUE THREAD

**Issue by lerogo** [tanhuajie/Reason-RFT#20]
Dear authors, thank you for your excellent work.
We can't reproduce the results of `GeoMath` and `Geometry3k`, and we need your help.
We conducted the experiment according to the script you provided, but we found that the results of reproducing `Task2` are significantly different from the paper.

Our Evaluation Results: [geometry3k_with_score.json](https://github.com/user-attachments/files/19729082/geometry3k_with_score.json 
 /  [geomath_with_score.json](https://github.com/user-attachments/files/19729083/geomath_with_score.json)
GeoMath(include choice & non-choice): 21.95
Geometry3k(include choice & non-choice): 25.88

Training settings: Task2、qwen2vl 2b、only_rl
Training script:
```bash
#!/bin/bash
timestamp=$1
# Dataset
# export TASK_NAME=geomath
export TASK_NAME=geoqa
export DATASET_NAME="tanhuajie2001/Reason-RFT-CoT-Dataset/train_jsons/A2-Structure-Perception-train-4k5-cot.json"
export IMAGE_PATH="tanhuajie2001/Reason-RFT-CoT-Dataset/images/train_images/"
export MODEL_NAME_OR_PATH="Qwen2-VL-2B-Instruct"
export OUTPUT_DIR="Reason-RFT/output/reason_rft_zero/${WANDB_RUN_NAME}"

# Debug
export DEBUG_MODE="True"
export LOG_PATH="${OUTPUT_DIR}/reward.log"

TORCH_GPUS=$(($GPUS - 1))

torchrun --master_addr ${MASTER_ADDR} --master-port ${MASTER_PORT} \
    --nnodes ${WORLD_SIZE} --node_rank ${RANK} --nproc-per-node=${TORCH_GPUS} \
  train/stage_rl/grpo.py \
  --deepspeed scripts/train/zero3.json \
  --output_dir ${OUTPUT_DIR} \
  --model_name_or_path ${MODEL_NAME_OR_PATH} \
  --dataset_name ${DATASET_NAME} \
  --image_path ${IMAGE_PATH} \
  --task_name ${TASK_NAME} \
  --pg_name ${pg_name} \
  --use_vllm_for_gen true \
  --use_system_prompt false \
  --max_prompt_length 4096 \
  --max_completion_length 512 \
  --num_generations 8 \
  --per_device_train_batch_size 1 \
  --gradient_accumulation_steps 2 \
  --logging_steps 1 \
  --bf16 \
  --report_to wandb \
  --gradient_checkpointing false \
  --attn_implementation flash_attention_2 \
  --max_pixels 480000 \
  --save_steps 100 \
  --num_train_epochs 1 \
  --run_name ${WANDB_RUN_NAME} \
  2>&1 | tee -a ${OUTPUT_DIR}/train.log
```

Training curve：https://wandb.lerogo.com/lerogo/vison-reason-rft/reports/qwen2vl_2b_task2_only_rl-20250412130441--VmlldzoxOQ?accessToken=t49yl3bjl1d8m1ybdlxky76e9evfio5kkxl01es4xviby5m24ab37vipm68j5un7

Evaluation script(`Reason-RFT/scripts/eval/open_source_models/single_gpu_eval/eval_by_vllm_task2_reason_rft_single_gpu.sh`):
```bash
DEVICE_ID=0
BATCH_SIZE=32

MODEL_NAME_OR_PATH="qwen2vl_2b_task2_only_rl-20250412130441"
BENCHMARK_LIST="geomath geometry3k"
STRATAGE_LIST="cot-sft cot-sft"

CUDA_VISIBLE_DEVICES=$DEVICE_ID python eval/eval_by_vllm_for_open_source.py \
    --batch_size $BATCH_SIZE \
    --model_name_or_path $MODEL_NAME_OR_PATH \
    --benchmark_list $BENCHMARK_LIST \
    --stratage_list $STRATAGE_LIST
```

Please note that we have modified the last line in the file `Reason-RFT/train/stage_rl/prompt.py` as follows:
```python
GEOQA_QUESTION_PROMPT = "{Question} Output the thinking process in <think> </think> and final answer (number or choice) in <answer> </answer> tags."
```

Thank you again for your work and looking forward to your reply.

**Comment by tanhuajie** [comment]
Thank you for your interest in our work. Based on the training script you provided, there are a few discrepancies we’d like to clarify:

1. You used `A2-Structure-Perception-train-4k5-cot.json` as the training sample. Has the sample format been converted as follows? Note the `"solution": "{answer}"`:
```json
{
    "id": "{id}",
    "image": "{image}",
    "problem": "{problem}",
    "solution": "{answer}"
},
```

2. I noticed that your script sets the number of epochs to 1 (`--num_train_epochs 1`), which differs from the setting in our paper (`--num_train_epochs 5`). This might be due to an issue with the training script I provided—my apologies. We will update it shortly. Additionally, I’d like to remind you that in our paper, the experiments for the `Structure Perception` task were trained for 5 epochs, while the other two tasks were trained for 1 epoch.

3. For your reference, we’ve also provided the training logs for this task:  
https://huggingface.co/tanhuajie2001/Reason-RFT-Zero-Structure-Perception-Qwen2-VL-2B/raw/main/train.log  

If you have any further questions, feel free to let me know.  🤗

**Comment by lerogo** [comment]
Thank you for your reply, I have modified the relevant files as you suggested. We will re-run the experiment and submit the updated results here. 
Additionally, it is worth mentioning that the replication of the other two tasks (RL training only) is fine, and can be replicated, but the results may have some fluctuations (hopefully this will be helpful to the community). 

Thanks again for your excellent work!

**Comment by lerogo** [comment]
Hi, 
after making modifications (changing epoch to 5), we still cannot reproduce the results of `GeoMath` and `Geometry3k`. 
We need your help.

Training curve: https://wandb.lerogo.com/lerogo/vison-reason-rft/reports/Run-qwen2vl_2b-7b_task2_only_rl--VmlldzozMA?accessToken=mklrisndwd2jdku6r48qeff2sm4n6r3hzq2l2a14ass1rsznyup7jinn4ip2qq1j

Eval results: [2b-vision-r1-result.zip](https://github.com/user-attachments/files/19787437/2b-vision-r1-result.zip)  [7b-vision-r1-result.zip](https://github.com/user-attachments/files/19787438/7b-vision-r1-result.zip)

**Comment by lerogo** [comment]
> Thank you for your interest in our work. Based on the training script you provided, there are a few discrepancies we’d like to clarify:
> 
> 1. You used `A2-Structure-Perception-train-4k5-cot.json` as the training sample. Has the sample format been converted as follows? Note the `"solution": "{answer}"`:
> 
> {
>     "id": "{id}",
>     "image": "{image}",
>     "problem": "{problem}",
>     "solution": "{answer}"
> },
> 2. I noticed that your script sets the number of epochs to 1 (`--num_train_epochs 1`), which differs from the setting in our paper (`--num_train_epochs 5`). This might be due to an issue with the training script I provided—my apologies. We will update it shortly. Additionally, I’d like to remind you that in our paper, the experiments for the `Structure Perception` task were trained for 5 epochs, while the other two tasks were trained for 1 epoch.
> 3. For your reference, we’ve also provided the training logs for this task:
>    https://huggingface.co/tanhuajie2001/Reason-RFT-Zero-Structure-Perception-Qwen2-VL-2B/raw/main/train.log
> 
> If you have any further questions, feel free to let me know. 🤗

We looked at this log and found that the training reward is similar to our curve. But our test results are very different.

**Comment by lerogo** [comment]
> Hi, after making modifications (changing epoch to 5), we still cannot reproduce the results of `GeoMath` and `Geometry3k`. We need your help.
> 
> Training curve: https://wandb.lerogo.com/lerogo/vison-reason-rft/reports/Run-qwen2vl_2b-7b_task2_only_rl--VmlldzozMA?accessToken=mklrisndwd2jdku6r48qeff2sm4n6r3hzq2l2a14ass1rsznyup7jinn4ip2qq1j
> 
> Eval results: [2b-vision-r1-result.zip](https://github.com/user-attachments/files/19787437/2b-vision-r1-result.zip) [7b-vision-r1-result.zip](https://github.com/user-attachments/files/19787438/7b-vision-r1-result.zip)
> 
> 2B: Average Score: 23.05, Count: 820 / Average Score: 11.25, Count: 800 7B: Average Score: 29.02, Count: 820 / Average Score: 42.12, Count: 800
> 
> More logs: [2b-train.log](https://github.com/user-attachments/files/19788783/2b-train.log) [7b-train.log](https://github.com/user-attachments/files/19788786/7b-train.log)
> 
> reward logs: https://drive.google.com/drive/folders/1Vuhgkiiz7sFQEsfyZ2bKsZifP7hxgjge


We also evaluated the model you released, and the evaluation results are as follows:

[2B](https://huggingface.co/tanhuajie2001/Reason-RFT-Zero-Structure-Perception-Qwen2-VL-2B)：
[ori2b-vision-r1-result.zip](https://github.com/user-attachments/files/19789163/ori2b-vision-r1-result.zip)
Average Score: **42.93**, Count: 820 / Average Score: 33.12, Count: 800

[7B](https://huggingface.co/tanhuajie2001/Reason-RFT-Zero-Structure-Perception-Qwen2-VL-7B)：
[ori7b-vision-r1-result.zip](https://github.com/user-attachments/files/19789164/ori7b-vision-r1-result.zip)
Average Score: 54.27, Count: 820 / Average Score: 55.25, Count: 800

**Comment by tanhuajie** [comment]
> > Hi, after making modifications (changing epoch to 5), we still cannot reproduce the results of `GeoMath` and `Geometry3k`. We need your help.
> > Training curve: https://wandb.lerogo.com/lerogo/vison-reason-rft/reports/Run-qwen2vl_2b-7b_task2_only_rl--VmlldzozMA?accessToken=mklrisndwd2jdku6r48qeff2sm4n6r3hzq2l2a14ass1rsznyup7jinn4ip2qq1j
> > Eval results: [2b-vision-r1-result.zip](https://github.com/user-attachments/files/19787437/2b-vision-r1-result.zip) [7b-vision-r1-result.zip](https://github.com/user-attachments/files/19787438/7b-vision-r1-result.zip)
> > 2B: Average Score: 23.05, Count: 820 / Average Score: 11.25, Count: 800 7B: Average Score: 29.02, Count: 820 / Average Score: 42.12, Count: 800
> > More logs: [2b-train.log](https://github.com/user-attachments/files/19788783/2b-train.log) [7b-train.log](https://github.com/user-attachments/files/19788786/7b-train.log)
> > reward logs: https://drive.google.com/drive/folders/1Vuhgkiiz7sFQEsfyZ2bKsZifP7hxgjge
> 
> We also evaluated the model you released, and the evaluation results are as follows:
> 
> [2B](https://huggingface.co/tanhuajie2001/Reason-RFT-Zero-Structure-Perception-Qwen2-VL-2B)： [ori2b-vision-r1-result.zip](https://github.com/user-attachments/files/19789163/ori2b-vision-r1-result.zip) Average Score: **42.93**, Count: 820 / Average Score: 33.12, Count: 800
> 
> [7B](https://huggingface.co/tanhuajie2001/Reason-RFT-Zero-Structure-Perception-Qwen2-VL-7B)： [ori7b-vision-r1-result.zip](https://github.com/user-attachments/files/19789164/ori7b-vision-r1-result.zip) Average Score: 54.27, Count: 820 / Average Score: 55.25, Count: 800
> 
> The results reported in the 7B model are not significantly different from those in the paper, but there is a significant difference in the 2B results.

Hi, @lerogo. Thank you for sharing your reproduction results. I just carefully compared the code we used for the experiments at the time with the currently open-sourced code and found the following **differences in the evaluation of `geomath`: During the evaluation, we used different prompts for choice questions and non-choice questions, whereas in the open-sourced version, we unified the prompts for both types of questions.** This might be the key reason.  

1. The shared prompt for choice and non-choice questions in the open-sourced version is:  
`"{Question}\nOutput the thinking process in <think> </think> and final answer (number or choice) in <answer> </answer> tags."  `

2. Our original prompt for choice questions was:  
`“{Question}\nPlease select the correct answer by writing the letter (A, B, C or D) that precedes your choice.\nOutput the thinking process in <think> </think> and final answer (chosen letter) in <answer> </answer> tags.”`

3. Our original prompt for non-choice questions was:  
`“{Question}\nOutput the thinking process in <think> </think> and final answer (float number or int number) in <answer> </answer> tags.”`

**Comment by tanhuajie** [comment]
> Hi, after making modifications (changing epoch to 5), we still cannot reproduce the results of `GeoMath` and `Geometry3k`. We need your help.
> 
> Training curve: https://wandb.lerogo.com/lerogo/vison-reason-rft/reports/Run-qwen2vl_2b-7b_task2_only_rl--VmlldzozMA?accessToken=mklrisndwd2jdku6r48qeff2sm4n6r3hzq2l2a14ass1rsznyup7jinn4ip2qq1j
> 
> Eval results: [2b-vision-r1-result.zip](https://github.com/user-attachments/files/19787437/2b-vision-r1-result.zip) [7b-vision-r1-result.zip](https://github.com/user-attachments/files/19787438/7b-vision-r1-result.zip)
> 
> 2B: Average Score: 23.05, Count: 820 / Average Score: 11.25, Count: 800 7B: Average Score: 29.02, Count: 820 / Average Score: 42.12, Count: 800
> 
> More logs: [2b-train.log](https://github.com/user-attachments/files/19788783/2b-train.log) [7b-train.log](https://github.com/user-attachments/files/19788786/7b-train.log)
> 
> reward logs: https://drive.google.com/drive/folders/1Vuhgkiiz7sFQEsfyZ2bKsZifP7hxgjge

I think this might also be an issue with the prompts. Please refer to the explanation above regarding the evaluation.

**Comment by tanhuajie** [comment]
Hi, @lerogo. Thanks for identifying the inconsistency in the prompts for task 2. we have fixed the issue accordingly. (PR #26)

**Comment by CYW06** [comment]
> Thank you for your interest in our work. Based on the training script you provided, there are a few discrepancies we’d like to clarify:
> 
> 1. You used `A2-Structure-Perception-train-4k5-cot.json` as the training sample. Has the sample format been converted as follows? Note the `"solution": "{answer}"`:
> 
> {
>     "id": "{id}",
>     "image": "{image}",
>     "problem": "{problem}",
>     "solution": "{answer}"
> },
> 2. I noticed that your script sets the number of epochs to 1 (`--num_train_epochs 1`), which differs from the setting in our paper (`--num_train_epochs 5`). This might be due to an issue with the training script I provided—my apologies. We will update it shortly. Additionally, I’d like to remind you that in our paper, the experiments for the `Structure Perception` task were trained for 5 epochs, while the other two tasks were trained for 1 epoch.
> 3. For your reference, we’ve also provided the training logs for this task:
>    https://huggingface.co/tanhuajie2001/Reason-RFT-Zero-Structure-Perception-Qwen2-VL-2B/raw/main/train.log
> 
> If you have any further questions, feel free to let me know. 🤗
你好 @tanhuajie RL-only实验中，每个json文件都要按上面的格式转换吗

**Comment by tanhuajie** [comment]
> > Thank you for your interest in our work. Based on the training script you provided, there are a few discrepancies we’d like to clarify:
> > 
> > 1. You used `A2-Structure-Perception-train-4k5-cot.json` as the training sample. Has the sample format been converted as follows? Note the `"solution": "{answer}"`:
> > 
> > {
> > "id": "{id}",
> > "image": "{image}",
> > "problem": "{problem}",
> > "solution": "{answer}"
> > },
> > 2. I noticed that your script sets the number of epochs to 1 (`--num_train_epochs 1`), which differs from the setting in our paper (`--num_train_epochs 5`). This might be due to an issue with the training script I provided—my apologies. We will update it shortly. Additionally, I’d like to remind you that in our paper, the experiments for the `Structure Perception` task were trained for 5 epochs, while the other two tasks were trained for 1 epoch.
> > 3. For your reference, we’ve also provided the training logs for this task:
> > https://huggingface.co/tanhuajie2001/Reason-RFT-Zero-Structure-Perception-Qwen2-VL-2B/raw/main/train.log
> > If you have any further questions, feel free to let me know. 🤗
> > 你好 [@tanhuajie](https://github.com/tanhuajie) RL-only实验中，每个json文件都要按上面的格式转换吗

是的，三个主任务分别对应A1，A2，A3的json数据集文件，需要根据readme的形式进行转换~

**Comment by Qistart** [comment]
> 是的，三个主任务分别对应A1，A2，A3的json数据集文件，需要根据readme的形式进行转换~

请问在项目中是否已提供相应的转换脚本（未找到🙃）
